missing cast between [*c]T and ?[*:0]T on fn parameter

First found while integrating with libexpat, but here's a test case:
```zig
const Handler = ?extern fn ([*c]const u8) void;
extern fn addCallback(handler: Handler) void;
fn myCallback(cstr: ?[*:0]const u8) callconv(.C) void {}

pub fn main() anyerror!void {
    addCallback(myCallback);
}
```

```
./test42.zig:8:17: error: expected type 'fn([*c]const u8) callconv(.C) void', found 'fn(?[*:0]const u8) callconv(.C) void'
    addCallback(myCallback);
                ^
./test42.zig:8:17: note: parameter 0: '?[*:0]const u8' cannot cast into '[*c]const u8'
    addCallback(myCallback);
                ^
./test42.zig:8:17: note: destination pointer requires a terminating '0' sentinel
    addCallback(myCallback);
                ^
```

```
< andrewrk> I just wanted to make sure that was there - so yes the compiler is incorrect; it should allow the cast. thanks
```

--- a/src/ir.cpp
+++ b/src/ir.cpp
@@ -11592,7 +11592,8 @@ static ConstCastOnly types_match_const_cast_only(IrAnalyze *ira, ZigType *wanted
             wanted_ptr_type->data.pointer.sentinel == nullptr ||
             (actual_ptr_type->data.pointer.sentinel != nullptr &&
              const_values_equal(ira->codegen, wanted_ptr_type->data.pointer.sentinel,
-                 actual_ptr_type->data.pointer.sentinel));
+                 actual_ptr_type->data.pointer.sentinel)) ||
+            actual_ptr_type->data.pointer.ptr_len == PtrLenC;
         if (!ok_null_term_ptrs) {
             result.id = ConstCastResultIdPtrSentinel;
             result.data.bad_ptr_sentinel = allocate_nonzero<ConstCastPtrSentinel>(1);

--- a/test/stage1/behavior/cast.zig
+++ b/test/stage1/behavior/cast.zig
@@ -768,3 +768,17 @@ test "variable initialization uses result locations properly with regards to the
     const x: i32 = if (b) 1 else 2;
     expect(x == 1);
 }
+
+test "cast between [*c]T and ?[*:0]T on fn parameter" {
+    const S = struct {
+        const Handler = ?extern fn ([*c]const u8) void;
+        fn addCallback(handler: Handler) void {}
+
+        fn myCallback(cstr: ?[*:0]const u8) callconv(.C) void {}
+
+        fn doTheTest() void {
+            addCallback(myCallback);
+        }
+    };
+    S.doTheTest();
+}

